### ISSUE TITLE
Considering Metalayer whole as a Model how do you train it? Or do I need to use the output of Metalayer in some other model and use it?

### ISSUE THREAD

**Issue by kameshso01** [rusty1s/pytorch_geometric#1008]
## ❓ Questions & Help
Hi, I am trying to use metalyer to give me a global update for graphs. But I am stuck at how do you pass the data to it? I am considering here Metalayer as a whole as a standalone model.

from torch.nn import Sequential
from torch import nn
from torch_scatter import scatter_mean
from torch_geometric.nn import MetaLayer

    class Edge_Model(torch.nn.Module):
    
    def __init__(self, num_node_feat, num_edge_feat, num_global_feat, hiddens):
        super(Edge_Model, self).__init__()
        in_channels = ( 2 * num_node_feat + num_edge_feat )
        self.edge_mlp = Sequential(
            nn.Linear(in_channels, hiddens),
            nn.ReLU(),
            nn.Linear(hiddens, 2*hiddens),
            nn.ReLU(),
            nn.Linear(2*hiddens, hiddens))
        
    def forward(self, src, dst, edge_attr, u, batch):
        out = torch.cat([src, dst, edge_attr, u[batch]], 1)
        return self.edge_mlp(out)   

    class Node_Model(torch.nn.Module):
    
    def __init__(self, num_node_feat, num_edge_feat, num_global_feat, hiddens):
        super(Node_Model, self).__init__()
        in_channels = num_node_feat + num_edge_feat
        self.node_mlp_1 = Sequential(
            nn.Linear(in_channels, hiddens),
            nn.ReLU(),
            nn.Linear(hiddens, 2*hiddens),
            nn.ReLU(),
            nn.Linear(2*hiddens, 2*hiddens + num_node_feat))
        self.node_mlp_2 = Sequential(
            nn.Linear(2*hiddens + num_node_feat, 2*hiddens),
            nn.ReLU(),
            nn.Linear(2*hiddens, in_channels))
        
    def forward(self, x, edge_index, edge_attr, u, batch):
        row, col = edge_index
        out = torch.cat([x[col], edge_attr], dim=1)
        out = self.node_mlp_1(out)
        out = scatter_mean(out, row, dim=0, dim_size=x.size(0))
        out = torch.cat([x, out], dim=1)
        return self.node_mlp_2(out)   

    class Global_Model(torch.nn.Module):
    
    def __init__(self, num_edge_feat, num_node_feat, num_global_feat, hiddens):
        super(Global_Model, self).__init__()
        in_channels = num_edge_feat + num_node_feat
        self.gloabal_mlp = Sequential(
            nn.Linear(in_channels, 2*hiddens),
            nn.ReLU(),
            nn.Linear(2*hiddens, 3*hiddens),
            nn.ReLU(),
            nn.Linear(3*hiddens, num_global_feat))
        
    def forward(self, x, edge_index, edge_attr, u, batch):
        out = torch.cat([u, scatter_mean(x, batch, dim=0)], dim=1)
        return self.gloabal_mlp(out)

    def build_meta_layer(num_edge_feat, num_node_feat, num_global_feat, hiddens):
         return MetaLayer(Edge_Model(num_edge_feat, num_node_feat, num_global_feat, hiddens),
                     Node_Model(num_edge_feat, num_node_feat, num_global_feat, hiddens),
                     Global_Model(num_edge_feat, num_node_feat, num_global_feat, hiddens))

       model = build_meta_layer(20,20,5,30)
       optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

       model.train()
       for batch in data_loader:
          optimizer.zero_grad()
          x, edge_attr, u = model(x=batch['x'], edge_attr=batch['edge_attr'], u=batch['u'], 
                            edge_index=batch['edge_index'], batch=batch['batch'])
       loss = F.nll_loss(u, batch['y'])
       loss.backward()
       optimizer.step()

**Comment by kameshso01** [comment]
![image](https://user-images.githubusercontent.com/13978944/75981580-cf45ef00-5f0a-11ea-8bc4-f5f647ab63bf.png)

**Comment by rusty1s** [comment]
This looks mostly correct to me. It seems that `cat` is failing since some of those tensors have only 1-dimensional tensors. You can verify that by `print(batch)`.

**Comment by kameshso01** [comment]
Thanks for Prompt reply. Yes. The global vector u  is tensor having five values which I want to be updated for ML against my labels which are the same as the shape of u.

**Comment by rusty1s** [comment]
Yeah, but if you concatenate, simply run `u.view(-1, 1)` beforehand.

**Comment by kameshso01** [comment]
Thanks. I will try and update.

**Comment by kameshso01** [comment]
Hi rusty1s, It seems that the error persists even if I run `u.view(-1,1)` beforehand. I am attaching an image of my batched data.





![p8KWprypZie8j5jKkK5f4JGTJaGlLMcz1GUGD5I7JDw original fullsize](https://user-images.githubusercontent.com/13978944/76078006-cbc46d80-5fc7-11ea-8aef-6e87fd3c9e39.png)

**Comment by rusty1s** [comment]
Well, this is a different error, right? Try setting `in_channels` in the global model to `in_channels = num_edge_feat + num_node_feat + num_global_feat`.

**Comment by kameshso01** [comment]
Hi, rusty1s I see that. But the error is still arising from edge_mlp from edge_model. I tried implementing edge_model as function like in issue #535. It works for [src, dst, edge_attr] whenever I introduce u in torch.cat function it give me this error. 

`class My_layer(nn.Module):
    
    def __init__(self, num_edge_feat, num_node_feat, num_glob_feat):
        super(My_layer, self).__init__()
        in_chan = 2 * num_edge_feat + num_node_feat
        self.edge_mlp = Sequential(
            nn.Linear(in_chan, num_node_feat),
            nn.ReLU())
    
        def edge_model(src, dst, edge_attr, u, batch):
            edge_neigh = torch.cat([src, dst, edge_attr, u[batch]],1)
            return self.edge_mlp(edge_neigh)
        
        self.op = MetaLayer(edge_model)
        
    def forward(self, x, edge_index, edge_attr, u, batch):
        return self.op(x, edge_index, edge_attr, batch)
    
class NET(nn.Module):
    
    def __init__(self, num_edge_feat, num_node_feat, num_glob_feat):
        super(NET, self).__init__()
        self.metalyer = My_layer(num_edge_feat, num_node_feat, num_glob_feat)
        
    def forward(self, data):
        x, edge_index, edge_attr, u, batch = data.x, data.edge_index, data.edge_attr, data.u, data.batch
        x, edge_attr, u = self.metalyer(x, edge_index, edge_attr, u, batch)
        return edge_attr

node_feat = dataset[0].x.shape[-1]
edge_feat = dataset[0].edge_attr.shape[-1]
glob_feat = dataset[0].u.shape[-1]

model = NET(node_feat, edge_feat, glob_feat)

optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

model.train()
for batch in data_loader:
    optimizer.zero_grad()
    output = model(batch)
    y = batch.y
    loss = torch.nn.functional.nll_loss(output, y)
    loss.backward()
    optimizer.step()`

**Comment by rusty1s** [comment]
There is a type issue in the list of concat tensors. You can check this, e.g., by printing out the type `print(xxx.dtype)`.

**Comment by kameshso01** [comment]
Hi, rusty1's are the updated edge_attr in the edge_model used in node_model? Like in graphnets where the aggregator function does that. Or meta-layer just outputs update edge_attr, x, and u.

**Comment by rusty1s** [comment]
Updated edge features will be used as edge features in the node model. We follow the GraphNet formulation here.

**Comment by kameshso01** [comment]
Hi rusty1s, Thanks for your help!
